### PR TITLE
migration: Add case for migration with shared TPM state

### PR DESCRIPTION
**Pull request by cliping**
VIRT-XXXX - Migrate vm with vtpm device with shared TPM state

### CODE REVIEW

**Comment by cliping** [comment]
Test result:
 (1/3) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_shared_tpm.nfs.persistent_and_p2p: PASS (259.21 s)
 (2/3) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_shared_tpm.nfs.persistent_and_non_p2p: PASS (319.66 s)
 (3/3) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_shared_tpm.nfs.transient_and_non_p2p: PASS (238.80 s)

 (1/3) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_shared_tpm.ceph.non_auth.persistent_and_p2p: PASS (722.23 s)
 (2/3) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_shared_tpm.ceph.non_auth.persistent_and_non_p2p: PASS (723.15 s)
 (3/3) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_shared_tpm.ceph.non_auth.transient_and_non_p2p: PASS (734.42 s)

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

File: libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+                   'ssh_remote_auth': True}
+    remote_virsh_session = virsh.VirshPersistent(**virsh_dargs)
+    libvirt_secret.clean_up_secrets(remote_virsh_session)
+    auth_sec_dict.update({"sec_uuid": src_tpm_sec_uuid})
+    params.update({"auth_sec_dict": auth_sec_dict})
+    dest_tmp_sec_uuid = libvirt.create_secret(auth_sec_dict, virsh_dargs)
+    remote_virsh_session.secret_set_value(dest_tmp_sec_uuid, secret_value, encode=True,
+                                          debug=True, ignore_status=True)
+    remote_virsh_session.close_session()
+    return src_tpm_sec_uuid
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py]: Removed return value.

File: libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    vm_session.close()
+    if remote:
+        if vm.serial_console is not None:
+            vm.cleanup_serial_console()
+        vm.connect_uri = src_uri
+    if cmd_result:
+        test.fail("Fail to run '%s': %s." % (tpm_cmd, cmd_result))
+
+
+def get_encryption_uuid(params):
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py]: Updated to 'set_secret()'. Thanks.

File: libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+    test.log.info("Setup vTPM device in guest xml.")
+    if set_remote_libvirtd_log:
+        migration_obj.set_remote_log()
+    if vm.is_alive():
+        vm.destroy(gracefully=False)
+    vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    # Remove all existing tpm devices
+    vmxml.remove_all_device_by_type('tpm')
+    vmxml.sync()
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py]: Updated.

File: libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+                   'remote_pwd': remote_pwd, 'unprivileged_user': None,
+                   'ssh_remote_auth': True}
+    remote_virsh_session = virsh.VirshPersistent(**virsh_dargs)
+    libvirt_secret.clean_up_secrets(remote_virsh_session)
+    auth_sec_dict.update({"sec_uuid": src_tpm_sec_uuid})
+    params.update({"auth_sec_dict": auth_sec_dict})
+    dest_tmp_sec_uuid = libvirt.create_secret(auth_sec_dict, virsh_dargs)
+    remote_virsh_session.secret_set_value(dest_tmp_sec_uuid, secret_value, encode=True,
+                                          debug=True, ignore_status=True)
+    remote_virsh_session.close_session()
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py]: Updated, thanks.

File: libvirt/tests/cfg/migration_with_vtpm/migration_with_shared_tpm.cfg [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            libvirtd_debug_level = "1"
+            libvirtd_debug_filters = "1:*"
+        - ceph:
+            disk_type = "network"
+            nfs_mount_dir =
+            mnt_path_name =
+            setup_local_nfs = 'no'
+            disk_source_protocol = "rbd"
+            disk_source_port = "6789"
+            disk_source_name = "migration-with-vtpm-pool/rbd.img"
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/cfg/migration_with_vtpm/migration_with_shared_tpm.cfg]: Updated, thanks.

**Review by dzhengfy — APPROVED** [review]
LGTM

File: libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        vm.destroy(gracefully=False)
+    vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    # Remove all existing tpm devices
+    vmxml.remove_all_device_by_type('tpm')
+
+    tpmxml = Tpm()
+    tpm_dict['backend']['encryption_secret'] = auth_sec_dict['sec_uuid']
+    tpmxml.setup_attrs(**tpm_dict)
+    vmxml.add_device(tpmxml)
+    vmxml.sync()
```
Comment (Yingshun) [inline libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py:159]: I would suggest using libvirt_vmxml.modify_vm_device() here, what do you think?

File: libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+    for disk in disks_dev:
+        vmxml.del_device(disk)
+    disk_xml = Disk(type_name="network")
+    disk_xml.setup_attrs(**disk_dict)
+    if auth_attrs:
+        auth_attrs = eval(auth_attrs)
+        auth_attrs['secret_uuid'] = auth_sec_dict['sec_uuid']
+        disk_xml.auth = disk_xml.new_auth(**auth_attrs)
+    disk_xml.source = disk_xml.new_disk_source(
+        **{"attrs": source_dict, "hosts": [host_dict]})
```
Comment (Yingshun) [inline libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py:237]: Can we update auth_attrs and source_dict to disk_dict and setup the disk with setup_attrs() in L231?

File: libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py (hunk 6 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py]: Updated, thanks.

File: libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py (hunk 7 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_vtpm/migration_with_shared_tpm.py]: Updated, thanks.

**Review by Yingshun — APPROVED** [review]

**Pull request merged**